### PR TITLE
Add metrics aspect to mixer.

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR adds the metrics aspect to mixer. This will allow the development of metrics backends that generate stats from Report() calls.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/159)

### CODE REVIEW

File: pkg/aspect/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	// reporting within the mixer.
+	Aspect interface {
+		aspect.Aspect
+
+		// Record directs a backend adapter to record the list of values
+		// that have been generated from Report() calls.
+		Record([]Value) error
+	}
+
+	// Value holds an single metric value that will be generated through
```
Comment (geeknoid) [inline pkg/aspect/metrics/metrics.go]: an -> a

**Comment by douglas-reid** [comment]
*[pkg/aspect/metrics/metrics.go, line 30 at r1](https://reviewable.io:443/reviews/istio/mixer/159#-KaIzB0rjM9iEIcwtHj-:-KaiGqdwtzjM3MXU5ih8:bxu03nd) ([raw file](https://github.com/istio/mixer/blob/a85b1023a9e4eba21d7f6a157c0e7790531fecc8/pkg/aspect/metrics/metrics.go#L30)):*
<details><summary><i>Previously, ZackButcher (Zack) wrote…</i></summary><blockquote>

> COUNTER is used to record increasing cumulative values.

Some metric backends allow decrementing counters (e.g. statsd), how do we plan to handle that case?
</blockquote></details>

At the moment, we don't have any plans for that.  I'd say we should address it when we compare schemas across metrics systems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159)*

**Comment by douglas-reid** [comment]
*[pkg/aspect/metrics/metrics.go, line 45 at r1](https://reviewable.io:443/reviews/istio/mixer/159#-KaIxJ3hUYEXEFQD8cF--r1-45:-KaiHI4pb9_X8-6TVbxP:b-896fix) ([raw file](https://github.com/istio/mixer/blob/a85b1023a9e4eba21d7f6a157c0e7790531fecc8/pkg/aspect/metrics/metrics.go#L45)):*
<details><summary><i>Previously, geeknoid (Martin Taillefer) wrote…</i></summary><blockquote>

an -> a
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159)*

**Comment by douglas-reid** [comment]
*[pkg/aspect/metrics/metrics.go, line 58 at r1](https://reviewable.io:443/reviews/istio/mixer/159#-KaIzhQPNQjadvOUvvaV:-KaiHJi_Bus066ss967o:baoh1iq) ([raw file](https://github.com/istio/mixer/blob/a85b1023a9e4eba21d7f6a157c0e7790531fecc8/pkg/aspect/metrics/metrics.go#L58)):*
<details><summary><i>Previously, ZackButcher (Zack) wrote…</i></summary><blockquote>

> // StringValue is used to pass a string-valued metric value.
> 		StringValue string
> 		// Int64Value is used to pass a integer-valued metric value.
> 		Int64Value int64
> 		// Float64Value is used to pass a double-valued metric value.
> 		Float64Value float64
> 		// BoolValue is used to pass a boolean-valued metric value.
> 		BoolValue bool

I don't like exposing the inner value fields publicly, I prefer providing a set of typesafe functions for handling the values in a metric. [opentracing's log package is very similar](https://github.com/opentracing/opentracing-go/blob/master/log/field.go).

I imagine something like:

    func StringValue(name string, kind Kind, labels map[string]interface{}, value string, start time.Time, end time.Time) Value {...}
    func Int64Value(name string, kind Kind, labels map[string]interface{}, value int64, start time.Time, end time.Time) Value {...}

    func (v Value) Extract() interface{} {
        return v.innerValue
    }
    
    func String(v Value) (string, err) {
       if val, ok := v.Extract().(string); ok {
          return val, nil
       }
       return nil, fmt.Errorf("Value is not a string")
    }

    func Int64(v Value) (int64, err) {
       if val, ok := v.Extract().(int64); ok {
          return val, nil
       }
       return nil, fmt.Errorf("Value is not an int64")
    }
</blockquote></details>

OK.  Add the methods as suggested to Value itself. Please take a look.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159)*

File: pkg/aspect/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		aspect.Adapter
+
+		// NewAspect returns a new quota implementation, based on the
+		// supplied Aspect configuration for the backend.
+		NewAspect(env aspect.Env, config proto.Message) (Aspect, error)
+	}
+)
+
+// String returns the string-valued metric value for a metrics.Value.
+func (v Value) String() (string, error) {
```
Comment (ZackButcher) [inline pkg/aspect/metrics/metrics.go]: To capture some of the talk in slack: upon thinking about this more, I'm not sure I'm convinced it's an improvement over making the user typecast themselves: we essentially changed the interface from

    s, ok := v.val.(string)
    if !ok { ...}

to

    s, err := v.String()
    if err != nil { ... }

However, this is easy enough to fix later, so I think we should proceed and we can reevaluate down the line.

**Comment by ZackButcher** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159)*

**Comment by ZackButcher** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159#-:-KaiQK2dGPYsyZ9B1x51:bnfp4nl)*

**Comment by douglas-reid** [comment]
*[pkg/aspect/metrics/metrics.go, line 86 at r2](https://reviewable.io:443/reviews/istio/mixer/159#-KaIxJ3hUYEXEFQD8cF--r2-86:-KaiU-aes8qyFR_si6Y4:bb6rskq) ([raw file](https://github.com/istio/mixer/blob/84227392b4cf5dbfb92dfbdf68d3eaa561bc791c/pkg/aspect/metrics/metrics.go#L86)):*
<details><summary><i>Previously, ZackButcher (Zack) wrote…</i></summary><blockquote>

To capture some of the talk in slack: upon thinking about this more, I'm not sure I'm convinced it's an improvement over making the user typecast themselves: we essentially changed the interface from

    s, ok := v.val.(string)
    if !ok { ...}

to

    s, err := v.String()
    if err != nil { ... }

However, this is easy enough to fix later, so I think we should proceed and we can reevaluate down the line.
</blockquote></details>

ACK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159)*

**Comment by geeknoid** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/159#-:-KaiV7IyEXm-sGGekzr3:bnfp4nl)*

**Pull request merged**